src/kinesis.ts

```diff
@@ -0,0 +1,157 @@
+import { LoggerInterface } from '@lifeomic/logging';
+import { v4 as uuid } from 'uuid';
+import { KinesisStreamEvent, Context as AWSContext } from 'aws-lambda';
+import {
+  BaseContext,
+  processWithOrdering,
+  withHealthCheckHandling,
+} from './utils';
+
+export type KinesisEventHandlerConfig<Event, Context> = {
+  /**
+   * A logger to use in the context.
+   */
+  logger: LoggerInterface;
+  /**
+   * A function for parsing the Kinesis event data into your custom type.
+   */
+  parseEvent: (body: string) => Event;
+  /**
+   * Create a "context" for the lambda execution. (e.g. "data sources")
+   */
+  createRunContext: (base: BaseContext) => Context | Promise<Context>;
+  /**
+   * The maximum concurrency for processing events.
+   *
+   * @default 5
+   */
+  concurrency?: number;
+};
+
+export type KinesisEventAction<Event, Context> = (
+  context: Context & BaseContext,
+  event: Event,
+) => void | Promise<void>;
+
+export type KinesisEventHandlerHarnessOptions<Event, Context> = {
+  /**
+   * A function for stringifying events.
+   */
+  stringifyEvent: (event: Event) => string;
+
+  /**
+   * An optional override for the logger.
+   */
+  logger?: LoggerInterface;
+
+  /**
+   * An optional override for creating the run context.
+   */
+  createRunContext?: () => Context | Promise<Context>;
+};
+
+export type KinesisEventHandlerHarnessContext<Event> = {
+  /** Sends the specified event through the handler. */
+  sendEvent: (event: { events: Event[] }) => Promise<void>;
+};
+
+/**
+ * An abstraction for a Kinesis event handler.
+ */
+export class KinesisEventHandler<Event, Context> {
+  private actions: KinesisEventAction<Event, Context>[] = [];
+
+  constructor(readonly config: KinesisEventHandlerConfig<Event, Context>) {}
+
+  /**
+   * Adds a event action to the handler.
+   *
+   * @param handler The handler, for additional chaining.
+   */
+  onEvent(
+    action: KinesisEventAction<Event, Context>,
+  ): KinesisEventHandler<Event, Context> {
+    this.actions.push(action);
+    return this;
+  }
+
+  lambda(): (event: KinesisStreamEvent, context: AWSContext) => Promise<void> {
+    return withHealthCheckHandling(async (event, awsContext) => {
+      // 1. Build the context.
+      const correlationId = uuid();
+      const context: BaseContext & Context = {
+        correlationId,
+        logger: this.config.logger.child({
+          requestID: awsContext.awsRequestId,
+          correlationId,
+        }),
+      } as any;
+
+      Object.assign(context, await this.config.createRunContext(context));
+
+      // 2. Process all the records.
+
+      await processWithOrdering(
+        {
+          items: event.Records,
+          orderBy: (record) => record.kinesis.partitionKey,
```

Comment:
I think this is the main place where things change when it comes to consuming events from Kinesis in Lambda. I think the `orderBy` is redundant in this scenario because all items in a batch already belong to the same shard (i.e. they all have the same partition key):

```
Lambda reads records from the data stream and invokes your function synchronously
with an event that contains stream records. Lambda reads records in batches and invokes 
your function to process records from the batch. Each batch contains records from a 
single shard/data stream.
```